File: src/main/java/frc/robot/subsystems/swerve/module/SwerveModule.java
package frc.robot.subsystems.swerve.module;

import edu.wpi.first.math.Vector;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.math.kinematics.SwerveModulePosition;
import edu.wpi.first.math.kinematics.SwerveModuleState;
import edu.wpi.first.math.numbers.N2;
import edu.wpi.first.math.system.plant.DCMotor;
import edu.wpi.first.units.Units;
import frc.robot.RobotType;
import java.util.function.Supplier;
import lib.math.VectorFunctions;
import lib.math.controllers.gains.ControllerGains;
import lib.math.controllers.gains.FeedforwardGains;
import lib.math.controllers.gains.PIDGains;
import org.littletonrobotics.junction.Logger;

/** Represents a single swerve module. */
public final class SwerveModule {
  private final int index;
  private final Translation2d modulePosition;
  private final ModuleIO io;
  private final ModuleIO.ModuleIOInputs inputs = new ModuleIO.ModuleIOInputs();
  private final Vector<N2> positiveRotationVector;

  private SwerveModuleState desiredState = new SwerveModuleState();
  private SwerveModuleState wheelForce = new SwerveModuleState();

  private static final double WHEEL_RADIUS_METERS = 0.049;
  private static final double TURN_GEARING = 150.0 / 7.0;
  private static final double DRIVE_GEARING = 6.75;

  private static final DCMotor DRIVE_MOTOR = DCMotor.getKrakenX60Foc(1);
  private static final DCMotor TURN_MOTOR = DCMotor.getNEO(1);

  private static final ControllerGains DRIVE_TORQUE_GAINS =
      new ControllerGains(new PIDGains(10.0, 0.0, 0.0), new FeedforwardGains(1.15, 0.0, 0.0));

  private static final ControllerGains DRIVE_VELOCITY_VOLTAGE_GAINS =
      new ControllerGains(
          new PIDGains(0.88842, 0.0, 0.0), new FeedforwardGains(0.34113, 1.9168, 0.081352));

  public SwerveModule(
      int driveID,
      int turnID,
      int encoderID,
      boolean invertDrive,
      boolean invertTurn,
      Rotation2d encoderOffset,
      Translation2d modulePosition) {
    this.index = (driveID - 1) / 3;
    this.modulePosition = modulePosition;

    positiveRotationVector =
        VectorFunctions.vectorFromRotation(modulePosition.getAngle().plus(Rotation2d.fromDegrees(90.0)));

    io = createIO(driveID, turnID, encoderID, invertDrive, invertTurn, encoderOffset);
  }

  private ModuleIO createIO(
      int driveID,
      int turnID,
      int encoderID,
      boolean invertDrive,
      boolean invertTurn,
      Rotation2d encoderOffset) {
    return switch (RobotType.MODE) {
      case SIMULATION ->
          new ModuleIOSim(
              new FeedforwardGains(0.0, 12.0 / 97.0, 0.0),
              new PIDGains(0.01, 0.0, 0.0),
              DRIVE_MOTOR,
              DRIVE_GEARING,
              new FeedforwardGains(),
              new PIDGains(),
              TURN_MOTOR,
              TURN_GEARING,
              WHEEL_RADIUS_METERS);
      case REAL ->
          new ModuleIOHybridFXS(
              driveID,
              DRIVE_GEARING,
              invertDrive,
              DRIVE_TORQUE_GAINS.getFeedforward(),
              DRIVE_TORQUE_GAINS.getPid(),
              turnID,
              TURN_GEARING,
              invertTurn,
              new FeedforwardGains(0.14403, 0.0, 0.0),
              new PIDGains(50.0, 0.0, 0.5),
              encoderID,
              encoderOffset,
              Units.Meters.of(WHEEL_RADIUS_METERS));
      case REPLAY -> new ModuleIO() {};
    };
  }

  public void periodic() {
    io.updateInputs(inputs);

    double driveError = inputs.driveVelocityMetersPerSec - desiredState.speedMetersPerSecond;
    Logger.recordOutput("modules/" + index + "/driveError", driveError);

    double turnError =
        desiredState.angle.minus(inputs.absoluteTurnPosition).getRadians();
    Logger.recordOutput("modules/" + index + "/turnError", turnError);
  }

  public void applyState(SwerveModuleState targetState) {
    SwerveModuleState optimized = SwerveModuleState.optimize(targetState, inputs.absoluteTurnPosition);
    SwerveModuleState scaled = SwerveModuleState.cosineScale(optimized, inputs.absoluteTurnPosition);

    Logger.recordOutput("modules/" + index + "/preOptimize", SwerveModuleState.struct, targetState);
    Logger.recordOutput("modules/" + index + "/postOptimize", SwerveModuleState.struct, scaled);

    io.setTurnPosition(scaled.angle);
    io.setDriveVelocity(scaled.speedMetersPerSecond);
    desiredState = scaled;
    wheelForce = new SwerveModuleState(0.0, scaled.angle);
  }

  public void applyState(SwerveModuleState targetState, Vector<N2> moduleForce) {
    SwerveModuleState optimized = SwerveModuleState.optimize(targetState, inputs.absoluteTurnPosition);
    SwerveModuleState scaled = SwerveModuleState.cosineScale(optimized, inputs.absoluteTurnPosition);

    Vector<N2> wheelDirection =
        VectorFunctions.vectorFromRotation(getState().angle);
    double wheelTorqueNm = moduleForce.dot(wheelDirection) * WHEEL_RADIUS_METERS;

    io.setTurnPosition(scaled.angle);
    io.setDriveVelocity(scaled.speedMetersPerSecond, wheelTorqueNm * DRIVE_MOTOR.KtNMPerAmp);

    wheelForce = new SwerveModuleState(wheelTorqueNm, scaled.angle);
    desiredState = scaled;
  }

  public void characterizeDriveVoltage(double volts) {
    io.setTurnPosition(Rotation2d.kZero);
    io.setDriveVoltage(volts);
  }

  public void characterizeSteerVoltage(double volts) {
    io.setTurnVoltage(volts);
    io.setDriveVoltage(0.0);
  }

  public void characterizeCurrent(double currentAmps) {
    io.setTurnPosition(Rotation2d.kZero);
    io.setDriveCurrent(currentAmps);
  }

  public SwerveModuleState getState() {
    return new SwerveModuleState(inputs.driveVelocityMetersPerSec, inputs.absoluteTurnPosition);
  }

  public SwerveModuleState getDesiredState() {
    return desiredState;
  }

  public SwerveModulePosition getPosition() {
    return new SwerveModulePosition(inputs.drivePositionMeters, inputs.absoluteTurnPosition);
  }

  public double getRadiusCharacterizationAngleRad() {
    return inputs.drivePositionRad;
  }

  public SwerveModuleState getWheelForce() {
    return wheelForce;
  }

  public Vector<N2> getPositiveRotationVector() {
    return positiveRotationVector;
  }

  public void stop() {
    io.stop();
  }

  public int getIndex() {
    return index;
  }

  public ModuleIO.ModuleIOInputs getInputs() {
    return inputs;
  }
}

File: src/main/java/frc/robot/subsystems/swerve/module/SwerveModule.kt
package frc.robot.subsystems.swerve.module

import edu.wpi.first.math.Vector
import edu.wpi.first.math.geometry.Rotation2d
import edu.wpi.first.math.geometry.Translation2d
import edu.wpi.first.math.kinematics.SwerveModulePosition
import edu.wpi.first.math.kinematics.SwerveModuleState
import edu.wpi.first.math.numbers.N2
import edu.wpi.first.math.system.plant.DCMotor
import edu.wpi.first.units.Units
import edu.wpi.first.units.Units.Amps
import edu.wpi.first.units.Units.Inches
import edu.wpi.first.units.Units.Meters
import edu.wpi.first.units.Units.Volts
import edu.wpi.first.units.measure.Angle
import edu.wpi.first.units.measure.Current
import edu.wpi.first.units.measure.Voltage
import frc.robot.RobotType
import lib.controllers.gains.ControllerGains
import lib.controllers.gains.FeedforwardGains
import lib.controllers.gains.PIDGains
import lib.math.units.into
import lib.math.units.measuredIn
import lib.math.vector
import org.littletonrobotics.junction.Logger

/**
 * Represents a swerve module on the robot.
 *
 * @param driveID The ID of the drive motor.
 * @param turnID The ID of the turn motor.
 * @param encoderID The ID of the encoder.
 * @param invertDrive Whether the drive motor is inverted.
 * @param invertTurn Whether the turn motor is inverted.
 * @param encoderOffset The offset of the encoder.
 * @param modulePosition The position of the module.
 */
class SwerveModule(
    val driveID: Int,
    val turnID: Int,
    val encoderID: Int,
    val invertDrive: Boolean,
    val invertTurn: Boolean,
    val encoderOffset: Rotation2d,
    val modulePosition: Translation2d
) {

    val driveMotor = DCMotor.getKrakenX60Foc(1)
    val turnMotor = DCMotor.getNEO(1)
    
    private val io: ModuleIO = when (RobotType.mode) {
        RobotType.Mode.SIMULATION -> ModuleIOSim(
            FeedforwardGains(
                kV = 12/97.0
            ),
            PIDGains(
                kP = 0.01
            ),
            driveMotor,
            8.14,
            FeedforwardGains(),
            PIDGains(),
            turnMotor,
            150 / 7.0,
            2.0 measuredIn Inches
        )
        RobotType.Mode.REAL -> ModuleIOTorqueCurrent(
            driveID,
            8.14, // 8.14:1 drive gearing
            invertDrive,
            driveTorqueGains.feedforward,
            driveTorqueGains.pid,
            turnID,
            150 / 7.0, // 150:7 turn gearing
            turnInverted = invertTurn,
            FeedforwardGains(kS = 0.14403),
            PIDGains(
                kP = 50.0,
                kD = 0.05
            ),
            encoderID,
            encoderOffset,
            wheelRadius = wheelRadius
        )
        else -> object : ModuleIO {}
    }
    
    /** The index of the module in the array of modules. Calculated with math */
    val index: Int = (driveID - 1) / 3

    /** The object that holds the inputs for the module. */
    val inputs: ModuleIO.ModuleInputs = ModuleIO.ModuleInputs()

    /** The state of the module. */
    val state: SwerveModuleState
        get() = SwerveModuleState(
            inputs.driveVelocity,
            inputs.absoluteTurnPosition
        )

    val radiusCharacterizationAngle: Angle
        get() = inputs.drivePositionAngular

    /** The position of the module. */
    val position: SwerveModulePosition
        get() = SwerveModulePosition(
            inputs.drivePosition,
            inputs.absoluteTurnPosition
        )
    
    var wheelForce: SwerveModuleState = SwerveModuleState()
        private set
    
    /** The desired state of the module. */
    var desiredState: SwerveModuleState = SwerveModuleState()

    /**
     * Vector pointing in the direction that the module would be facing if it was contributing 100% of its velocity
     * to the robot rotating counter-clockwise.
     */
    val positiveRotationVector = (modulePosition.angle + Rotation2d.fromDegrees(90.0)).vector

    /** Updates the inputs for the module. */
    fun periodic() {
        io.updateInputs(inputs)
        
        Logger.recordOutput("modules/$index/driveError", inputs.driveVelocity.minus(desiredState.speedMetersPerSecond measuredIn Units.MetersPerSecond))
        Logger.recordOutput("modules/$index/turnError", inputs.absoluteTurnPosition.minus(desiredState.angle))
    }

    /**
     * Applies the desired state to the module.
     *
     * @param desiredState The desired state for the module.
     */
    fun applyState(desiredState: SwerveModuleState) {
        desiredState.optimize(inputs.absoluteTurnPosition)
        desiredState.cosineScale(inputs.absoluteTurnPosition)

        io.setTurnPosition(desiredState.angle)
        io.setDriveVelocity(desiredState.speedMetersPerSecond measuredIn Units.MetersPerSecond)
        this.desiredState = desiredState
    }
    
    fun applyState(desiredState: SwerveModuleState, wheelForce: Vector<N2>) {
        desiredState.optimize(inputs.absoluteTurnPosition)
        desiredState.cosineScale(inputs.absoluteTurnPosition)
        
        val wheelDirection = state.angle.vector
        
        val wheelTorqueNm = wheelForce.dot(wheelDirection) * (wheelRadius into Meters)

        io.setTurnPosition(desiredState.angle)
        io.setDriveVelocity(
            desiredState.speedMetersPerSecond measuredIn Units.MetersPerSecond,
            (wheelTorqueNm * driveMotor.KtNMPerAmp) measuredIn Amps)
        
        this.wheelForce = SwerveModuleState(
            wheelTorqueNm,
            desiredState.angle
        )
        this.desiredState = desiredState
    }

    fun characterizeDriveVoltage(volts: Voltage) {
        io.setTurnPosition(Rotation2d())
        io.setDriveVoltage(volts)
    }

    fun characterizeSteerVoltage(volts: Voltage) {
        io.setTurnVoltage(volts)
        io.setDriveVoltage(Volts.zero())
    }

    fun characterizeCurrent(current: Current){
        io.setTurnPosition(Rotation2d())
        io.setDriveCurrent(current)
    }

    private companion object {
        val wheelRadius = 1.947 measuredIn Inches

        val driveTorqueGains: ControllerGains = ControllerGains(
            PIDGains(
                kP = 10.0
            ),
            FeedforwardGains(
                kS = 1.15
            )
        )

        val driveVelocityVoltageGains: ControllerGains = ControllerGains(
            PIDGains(
                kP = 0.88842,
            ),
            FeedforwardGains(
                kV = 1.9168,
                kS = 0.34113,
                kA = 0.081352
            )
        )

    }
}